File: app/src/main/java.backup/com/lumiyaviewer/lumiya/slproto/messages/RegionPresenceResponse.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.google.common.primitives.UnsignedBytes;
import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.net.Inet4Address;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.UUID;

public class RegionPresenceResponse extends SLMessage {
    public ArrayList<RegionData> RegionData_Fields = new ArrayList<>();

    public static class RegionData {
        public Inet4Address ExternalRegionIP;
        public Inet4Address InternalRegionIP;
        public byte[] Message;
        public long RegionHandle;
        public UUID RegionID;
        public int RegionPort;
        public double ValidUntil;
    }

    public RegionPresenceResponse() {
        this.zeroCoded = true;
    }

    public int CalcPayloadSize() {
        int i = 5;
        Iterator<T> it = this.RegionData_Fields.iterator();
        while (true) {
            int i2 = i;
            if (!it.hasNext()) {
                return i2;
            }
            i = ((RegionData) it.next()).Message.length + 43 + i2;
        }
    }

    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleRegionPresenceResponse(this);
    }

    public void PackPayload(ByteBuffer byteBuffer) {
        byteBuffer.putShort(-1);
        byteBuffer.put((byte) 0);
        byteBuffer.put((byte) 16);
        byteBuffer.put((byte) this.RegionData_Fields.size());
        for (RegionData regionData : this.RegionData_Fields) {
            packUUID(byteBuffer, regionData.RegionID);
            packLong(byteBuffer, regionData.RegionHandle);
            packIPAddress(byteBuffer, regionData.InternalRegionIP);
            packIPAddress(byteBuffer, regionData.ExternalRegionIP);
            packShort(byteBuffer, (short) regionData.RegionPort);
            packDouble(byteBuffer, regionData.ValidUntil);
            packVariable(byteBuffer, regionData.Message, 1);
        }
    }

    public void UnpackPayload(ByteBuffer byteBuffer) {
        byte b = byteBuffer.get() & UnsignedBytes.MAX_VALUE;
        for (int i = 0; i < b; i++) {
            RegionData regionData = new RegionData();
            regionData.RegionID = unpackUUID(byteBuffer);
            regionData.RegionHandle = unpackLong(byteBuffer);
            regionData.InternalRegionIP = unpackIPAddress(byteBuffer);
            regionData.ExternalRegionIP = unpackIPAddress(byteBuffer);
            regionData.RegionPort = unpackShort(byteBuffer) & 65535;
            regionData.ValidUntil = unpackDouble(byteBuffer);
            regionData.Message = unpackVariable(byteBuffer, 1);
            this.RegionData_Fields.add(regionData);
        }
    }
}

File: Linkpoint/src/main/kotlin/com/linkpoint/slproto/messages/RegionPresenceResponse.kt
package com.linkpoint.slproto.messages

import com.google.common.primitives.UnsignedBytes
import com.linkpoint.slproto.SLMessage
import java.net.Inet4Address
import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.Iterator
import java.util.UUID

class RegionPresenceResponse : SLMessage() {
    public ArrayList<RegionData> RegionData_Fields = ArrayList<>()

    @JvmStatic
    class RegionData {
        public Inet4Address ExternalRegionIP
        public Inet4Address InternalRegionIP
        public ByteArray Message
        public Long RegionHandle
        public UUID RegionID
        public Int RegionPort
        public Double ValidUntil
    }

    public RegionPresenceResponse() {
        this.zeroCoded = true
    }

    public fun CalcPayloadSize(): Int {
        val i: Int = 5
        val it: Iterator<T> = this.RegionData_Fields.iterator()
        while (true) {
            val i2: Int = i
            if (!it.hasNext()) {
                return i2
            }
            i = ((RegionData) it.next()).Message.length + 43 + i2
        }
    }

    fun Handle(sLMessageHandler: SLMessageHandler) {
        sLMessageHandler.HandleRegionPresenceResponse(this)
    }

    fun PackPayload(byteBuffer: ByteBuffer) {
        byteBuffer.putShort(-1)
        byteBuffer.put((Byte) 0)
        byteBuffer.put((Byte) 16)
        byteBuffer.put((Byte) this.RegionData_Fields.size())
        for (RegionData regionData : this.RegionData_Fields) {
            packUUID(byteBuffer, regionData.RegionID)
            packLong(byteBuffer, regionData.RegionHandle)
            packIPAddress(byteBuffer, regionData.InternalRegionIP)
            packIPAddress(byteBuffer, regionData.ExternalRegionIP)
            packShort(byteBuffer, (Short) regionData.RegionPort)
            packDouble(byteBuffer, regionData.ValidUntil)
            packVariable(byteBuffer, regionData.Message, 1)
        }
    }

    fun UnpackPayload(byteBuffer: ByteBuffer) {
        val b: Byte = byteBuffer.get() & UnsignedBytes.MAX_VALUE
        for (Int i = 0; i < b; i++) {
            val regionData: RegionData = RegionData()
            regionData.RegionID = unpackUUID(byteBuffer)
            regionData.RegionHandle = unpackLong(byteBuffer)
            regionData.InternalRegionIP = unpackIPAddress(byteBuffer)
            regionData.ExternalRegionIP = unpackIPAddress(byteBuffer)
            regionData.RegionPort = unpackShort(byteBuffer) & 65535
            regionData.ValidUntil = unpackDouble(byteBuffer)
            regionData.Message = unpackVariable(byteBuffer, 1)
            this.RegionData_Fields.add(regionData)
        }
    }
}
